js/logic.js

```diff
@@ -0,0 +1,862 @@
+            } else {
+              if (this.failOnMismatchedDetails) {
+                removeMatchListener(event);
+                reject(new MismatchError(this, event));
+                return true; // matched
+              } else {
+                // Ignore mismatched events; maybe we'll match later.
+              }
+            }
+            return false; // not done yet, didn't find a match
+          };
+
+          emitter.on('*', matchFn);
+
+          if (opts.prevMatcher) {
+            // Run the match function on missed logs, until one of
+            // them returns true (indicating we matched and are done).
+            opts.prevMatcher.capturedLogs.some(matchFn);
+          }
+
+        }, (error) => {
+          // Pass along any existing errors.
+          reject(error);
+        });
+      });
+    } else {
+      // This is the '.then()' case; we still want to return a
+      // LogicMatcher so they can chain, but without any further expectations.
+      this.promise = prevPromise;
+    }
+  }
+
+  LogicMatcher.normalizeMatchArgs = function(ns, type, details) {
+    // 'ns' is optional
+    if (typeof type === 'object') {
+      details = type;
+      type = ns;
+      ns = null;
+    }
+    return { ns: ns, type: type, detailPredicate: details };
+  }
+
+  LogicMatcher.prototype = {
+
+    /**
+     * Same as `logic.match`.
+     */
+    match(ns, type, details) {
+      var args = LogicMatcher.normalizeMatchArgs(ns, type, details);
+      args.prevMatcher = this;
+      args.prevPromise = this.promise;
+      return new LogicMatcher(args);
+    },
+
+    /**
+     * Look at THE LOGS ALREADY CAPTURED by this LogicMatcher, and fail if any
+     * of them match this one.
+     */
+    failIfMatched(ns, type, details) {
+      var args = LogicMatcher.normalizeMatchArgs(ns, type, details);
+      args.not = true;
+      args.prevMatcher = this;
+      args.prevPromise = this.promise;
+      return new LogicMatcher(args);
+    },
+
+    /**
+     * Like Promise.then(); resolves with an array of matched logs.
+     */
+    then(fn, catchFn) {
+      return new LogicMatcher({
+        prevPromise: this.promise.then(() => {
+          var ret = fn(this.matchedLogs.slice());
+          if (ret instanceof Promise) {
+            ret = new LogicMatcher({
+              prevPromise: ret
+            });
+          }
+          return ret;
+        }, catchFn)
+      });
+    },
+
+    toString() {
+      return '<LogicMatcher ' + (this.ns ? this.ns + '/' : '') +
+        this.type + ' ' + new ObjectSimplifier().simplify(this.detailPredicate)
+        + '>';
+    }
+  }
+
+  function Scope(namespace, defaultDetails) {
+    this.namespace = namespace;
+
+    if (defaultDetails && !isPlainObject(defaultDetails)) {
+      throw new Error('Invalid defaultDetails; expected a plain-old object: ' +
+                      defaultDetails);
+    }
+    this.defaultDetails = defaultDetails;
+  }
+
+  function ObjectSimplifier(opts) {
```

Comment:
This is great.  I'm pretty sure circular buffer saving has been broken for a while because we'd inevitably put something in that JSON.stringify would explode on and I think the platform might have gotten more strict at some point.  Not sure what was breaking us.  (Probably blobs or typed arrays/array buffers or something.)